### PR TITLE
Add new timestamp formats and refactor timestamp code in clp-s (fixes #261)

### LINKED ISSUE #261: clp-s: Nested timestamps result in incorrect mst node names
### Bug

Ingesting records with nested timestamps using the `--timestamp-key` argument turns the key name in the last level of the MST into the fully qualified name. e.g. records like `{"a": "ts": ...}` will look like `{"a": "a.ts": ...}` when decompressed.

### CLP version

e80cee00c38f0e403e24ab2f3475ef03b0bc0604

### Environment

Ubuntu focal clp image

### Reproduction steps

1) Ingest a dataset with nested timestamps while using the `--timestamp-key` argument
2) Observe incorrect data upon decompression

### PR DESCRIPTION
**Pull request by gibber9809**
# References
Fixes #261 

# Description
* Add some obvious variations of existing timestamp patterns to clp, clp-s, and glt
* Add "%E" format specifier to clp-s to capture unix epoch millisecond timestamps found in strings
* Refactor and simplify timestamp column compression code in clp-s
* Fix a nested timestamp bug

# Validation performed
* Compressed data with the new timestamp formats and verified that the timestamps were parsed succesfully
* Verified that nested timestamps now exhibit expected behaviour for compression/decompression/search
* Added unit tests in clp for new timestamp formats
* Ingested [cockroachdb](https://zenodo.org/records/10516387) dataset and performed searches on timestamp column

### CODE REVIEW

**Review by wraymo — CHANGES_REQUESTED** [review]
Great work! Just did the first pass. One thing I'm little bit concerned about is if we mix timestamps in millisecond with timestamp in nanoseconds, we can get wrong results during search. For example, if we have two log messages `{"timestamp": "1707026480000"}` and `{"timestamp": "1707026455.123456789"}`, and the query is `timestamp > date(1707026480000)`, we should return nothing but actually we will return the second log message.

**Comment by gibber9809** [comment]
> Great work! Just did the first pass. One thing I'm little bit concerned about is if we mix timestamps in milliseconds with timestamp in nanoseconds, we can get wrong results during search. For example, if we have two log messages `{"timestamp": "1707026480000"}` and `{"timestamp": "1707026455.123456789"}`, and the query is `timestamp > date(1707026480000)`, we should return nothing but actually we will return the second log message.

Right this is a good point. I modified how we parse literals supplied inside of `date()` so that its at least consistent with how we parse the log. E.g. if we have timestamps like `"1707026455.123456789"` in the log and you copy and paste timestamps to specify a filter like `timestamp: date(1707026455.123456789)` then the timestamp will match as expected. You are right though that when the user starts mixing formats they can get unexpected results.

The core of the problem is that our various timestamp formats already implicitly have different levels of precision (second, millisecond, nanosecond) but we haven't really reconciled these different precisions at the storage, search, or archive metadata levels.

We sort of deal with this problem halfway in TimestampPattern as timestamps which specify second-level precision get stored with millisecond precision, but the same can't be said for raw integer or floating point timestamps. One way to handle this might be to transform all timestamps into nanosecond precision timestamps no matter how they're parsed -- this would make storage, search, and archive metadata consistent in an easy to reason about way. The downside is that this creates an encoding challenge for raw int and float timestamp columns if we want to decompress in a 1:1 way. 

The other issue is that "store everything as nanosecond precision" is a bit loaded because it might sometimes be ambiguous/hard to figure out what the precision of some timestamp is supposed to be, so there's still a need to let users disambiguate via configuration for a robust solution (granted for most timestamps it should be pretty obvious).

Anyway if we do something like this (which we should at some point) I'm not sure if we want to include it in this PR (even though its slowly becoming the fix-every-timestamp-issue-we-notice PR).

**Comment by wraymo** [comment]
And can we delete `TimestampEntry::TimestampEncoding` and associated `ingest_timestamp(double)`?

**Comment by gibber9809** [comment]
> And can we delete `TimestampEntry::TimestampEncoding` and associated `ingest_timestamp(double)`?

Not yet because we still allow raw floating point columns to be timestamps.

**Comment by wraymo** [comment]
> > And can we delete `TimestampEntry::TimestampEncoding` and associated `ingest_timestamp(double)`?
> 
> Not yet because we still allow raw floating point columns to be timestamps.

Oh, that's right

**Comment by wraymo** [comment]
> > Great work! Just did the first pass. One thing I'm little bit concerned about is if we mix timestamps in milliseconds with timestamp in nanoseconds, we can get wrong results during search. For example, if we have two log messages `{"timestamp": "1707026480000"}` and `{"timestamp": "1707026455.123456789"}`, and the query is `timestamp > date(1707026480000)`, we should return nothing but actually we will return the second log message.
> 
> Right this is a good point. I modified how we parse literals supplied inside of `date()` so that its at least consistent with how we parse the log. E.g. if we have timestamps like `"1707026455.123456789"` in the log and you copy and paste timestamps to specify a filter like `timestamp: date(1707026455.123456789)` then the timestamp will match as expected. You are right though that when the user starts mixing formats they can get unexpected results.
> 
> The core of the problem is that our various timestamp formats already implicitly have different levels of precision (second, millisecond, nanosecond) but we haven't really reconciled these different precisions at the storage, search, or archive metadata levels.
> 
> We sort of deal with this problem halfway in TimestampPattern as timestamps which specify second-level precision get stored with millisecond precision, but the same can't be said for raw integer or floating point timestamps. One way to handle this might be to transform all timestamps into nanosecond precision timestamps no matter how they're parsed -- this would make storage, search, and archive metadata consistent in an easy to reason about way. The downside is that this creates an encoding challenge for raw int and float timestamp columns if we want to decompress in a 1:1 way.
> 
> The other issue is that "store everything as nanosecond precision" is a bit loaded because it might sometimes be ambiguous/hard to figure out what the precision of some timestamp is supposed to be, so there's still a need to let users disambiguate via configuration for a robust solution (granted for most timestamps it should be pretty obvious).
> 
> Anyway if we do something like this (which we should at some point) I'm not sure if we want to include it in this PR (even though its slowly becoming the fix-every-timestamp-issue-we-notice PR).

Not sure if I'm right. From what I've observed, only `%F` represents a nanosecond timestamp. All other formats are converted into milliseconds. So maybe we can just handle this specific nanosecond case?

File: components/core/src/clp_s/ColumnWriter.hpp [hunk 1]
```diff
@@ -188,54 +174,20 @@ class VariableStringColumnWriter : public BaseColumnWriter {
...
 public:
     // Constructor
-    DateStringColumnWriter(
-            std::string const& name,
-            int32_t id,
-            std::shared_ptr<TimestampDictionaryWriter> timestamp_dict
-    )
-            : BaseColumnWriter(name, id),
-              m_timestamp_dict(std::move(timestamp_dict)) {}
+    DateStringColumnWriter(int32_t id) : BaseColumnWriter(id) {}
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/ColumnWriter.hpp:177]:
```suggestion
    explicit DateStringColumnWriter(int32_t id) : BaseColumnWriter(id) {}
```

File: components/core/src/clp_s/TimestampPattern.hpp [hunk 2]
```diff
@@ -43,6 +43,8 @@ namespace clp_s {
  * - S  2-digit 0-padded second (00-60) (60 to account for leap seconds)
  * - 3  0-padded millisecond (000-999)
  * - T  0-padded millisecond no trailing 0 (000)-999) e.g. (000), 9(00), 99(0), 099
+ * - E  N-digit UNIX epoch timestamp in milliseconds
+ * - F  N-digit UNIX epoch timestamp in floating point format with 9-digits trailing the decimal
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.hpp:47]:
```suggestion
 * - E  N-digit millisecond-precision UNIX epoch timestamp
 * - F  N-digit nanosecond-precision UNIX epoch timestamp in floating-point format with 9 digits
 *      trailing the decimal
```

I'm not sure "N-digit" adds much, but I don't mind if you want to leave it in.

File: components/core/src/clp_s/TimestampPattern.cpp [hunk 3]
```diff
@@ -772,6 +799,57 @@ bool TimestampPattern::parse_timestamp(
                     break;
                 }
 
+                case 'E': {  // UNIX epoch milliseconds timestamp
+                    // only allow consuming entire timestamp string
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.cpp:803]:
```suggestion
                case 'E': {  // Millisecond-precision UNIX epoch timestamp
                    // Only allow consuming entire timestamp string
```

File: components/core/src/clp_s/TimestampPattern.cpp [hunk 4]
```diff
@@ -772,6 +799,57 @@ bool TimestampPattern::parse_timestamp(
...
+                    // epoch seconds and nanoseconds as well)
+                    if (line_ix > 0 || false == convert_string_to_int(line, timestamp)) {
+                        return false;
+                    }
+                    timestamp_begin_pos = 0;
+                    timestamp_end_pos = line.length();
+                    return true;
+                }
+
+                case 'F': {  // Nanosecond precision floating point UNIX epoch timestamp
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.cpp:815]:
```suggestion
                case 'F': {  // Nanosecond-precision floating-point UNIX epoch timestamp
```

File: components/core/src/clp_s/TimestampPattern.cpp [hunk 5]
```diff
@@ -772,6 +799,57 @@ bool TimestampPattern::parse_timestamp(
...
+                    constexpr auto cNanosecondDigits = 9;
+                    constexpr auto cNanosecondMultiplier = 1'000'000'000;
+                    // only allow consuming entire timestamp string
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.cpp:818]:
```suggestion
                    // Only allow consuming entire timestamp string
```

File: components/core/src/clp_s/TimestampPattern.cpp [hunk 6]
```diff
@@ -772,6 +799,57 @@ bool TimestampPattern::parse_timestamp(
...
+                    if (line_ix > 0) {
+                        return false;
+                    }
+                    auto dot_position = line.find('.');
+                    auto nanosecond_start = dot_position + 1;
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.cpp:823]: `nanoseconds_begin_pos`?

File: components/core/src/clp_s/TimestampPattern.cpp [hunk 7]
```diff
@@ -772,6 +799,57 @@ bool TimestampPattern::parse_timestamp(
...
+                    if (std::string::npos == dot_position || dot_position == 0
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/TimestampPattern.cpp:824]:
```suggestion
                    if (std::string::npos == dot_position || 0 == dot_position
```

File: components/core/src/clp_s/TimestampPattern.hpp (hunk 2 above)
Comment (kirkrodrigues) [reply components/core/src/clp_s/TimestampPattern.hpp:47]: While we're in the area, does the description for `T` have an extra `)` in it?

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
Deferring to @wraymo for the validity of the clp-s timestamp-related changes. Added some comments about the new formats added and minor style changes.

File: components/core/src/clp_s/TimestampPattern.hpp (hunk 2 above)
Comment (gibber9809) [reply components/core/src/clp_s/TimestampPattern.hpp:47]: yep, I can take that out

File: components/core/src/clp_s/TimestampPattern.hpp [hunk 8]
```diff
@@ -43,6 +43,8 @@ namespace clp_s {
  * - S  2-digit 0-padded second (00-60) (60 to account for leap seconds)
  * - 3  0-padded millisecond (000-999)
  * - T  0-padded millisecond no trailing 0 (000)-999) e.g. (000), 9(00), 99(0), 099
```
Comment (gibber9809) [inline components/core/src/clp_s/TimestampPattern.hpp:45]:
```suggestion
 * - T  0-padded millisecond no trailing 0 (000-999) e.g. (000), 9(00), 99(0), 099
```

File: components/core/src/glt/TimestampPattern.cpp [hunk 9]
```diff
@@ -142,6 +142,18 @@ void TimestampPattern::init() {
...
+    // E.g. 2015/01/31 15:50:45,123
+    patterns.emplace_back(0, "%Y/%m/%d %H:%M:%S,%3");
+    // E.g. 2015/01/31T15:50:45
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S");
+    // E.g. 2015/01/31T15:50:45.123
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S.%3");
+    // E.g. 2015/01/31T15:50:45,123
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S,%3");
+    // E.g. 15/01/31 15:50:45
+    patterns.emplace_back(0, "%y/%m/%d %H:%M:%S");
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/glt/TimestampPattern.cpp:156]: good catch

File: components/core/src/clp/TimestampPattern.cpp [hunk 10]
```diff
@@ -140,6 +140,16 @@ void TimestampPattern::init() {
...
+    // E.g. 2015/01/31 15:50:45.123
+    patterns.emplace_back(0, "%Y/%m/%d %H:%M:%S.%3");
+    // E.g. 2015/01/31 15:50:45,123
+    patterns.emplace_back(0, "%Y/%m/%d %H:%M:%S,%3");
+    // E.g. 2015/01/31T15:50:45
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S");
+    // E.g. 2015/01/31T15:50:45.123
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S.%3");
+    // E.g. 2015/01/31T15:50:45,123
+    patterns.emplace_back(0, "%Y/%m/%dT%H:%M:%S,%3");
```
[the remark this answers is not in the document]
Comment (gibber9809) [reply components/core/src/clp/TimestampPattern.cpp:152]: Sure, should I sort all of the timestamp formats in the file? I don't think they're consistently sorted with millisecond precision timestamps first.

[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/clp/TimestampPattern.cpp:152]: Sure (I think they are sorted such that where two patterns have a common prefix, the more precise version appears first, but it is clearer to group all the more precise timestamps together before the less precise versions).

**Review by kirkrodrigues — APPROVED** [review]
Commit msg: `Add new timestamp formats and simplify timestamp column compression in clp-s (fixes #261).`?

Although we should include some text about what was the fix for #261.

**Review by wraymo — APPROVED** [review]

**Pull request merged**